README.md
# servant-pagination [![](https://img.shields.io/hackage/v/servant-pagination.svg)](https://hackage.haskell.org/package/servant-pagination) [![Build Status](https://travis-ci.com/chordify/haskell-servant-pagination.svg?branch=master)](https://app.travis-ci.com/github/chordify/haskell-servant-pagination)

## Overview

This module offers opinionated helpers to declare a type-safe and a flexible pagination
mechanism for Servant APIs. This design, inspired by [Heroku's API](https://devcenter.heroku.com/articles/platform-api-reference#ranges),
provides a small framework to communicate about a possible pagination feature of an endpoint,
enabling a client to consume the API in different fashions (pagination with offset / limit,
endless scroll using last referenced resources, ascending and descending ordering, etc.)

Therefore, client can provide a `Range` header with their request with the following format:

- `Range: <field> [<value>][; offset <o>][; limit <l>][; order <asc|desc>]`

For example: `Range: createdAt 2017-01-15T23%3A14%3A67.000Z; offset 5; order desc` indicates that
the client is willing to retrieve the next batch of document in descending order that were
created after the fifteenth of January, skipping the first 5.

As a response, the server may return the list of corresponding document, and augment the
response with 3 headers:

- `Accept-Ranges`: A comma-separated list of fields upon which a range can be defined
- `Content-Range`: Actual range corresponding to the content being returned
- `Next-Range`: Indicate what should be the next `Range` header in order to retrieve the next range

For example:

- `Accept-Ranges: createdAt, modifiedAt`
- `Content-Range: createdAt 2017-01-15T23%3A14%3A51.000Z..2017-02-18T06%3A10%3A23.000Z`
- `Next-Range: createdAt 2017-02-19T12%3A56%3A28.000Z; offset 0; limit 100; order desc`


## Getting Started

Code-wise the integration is quite seamless and unobtrusive. `servant-pagination` provides a
`Ranges (fields :: [Symbol]) (resource :: *) -> *` data-type for declaring available ranges
on a group of _fields_ and a target _resource_. To each combination (resource + field) is
associated a given type `RangeType (resource :: *) (field :: Symbol) -> *` as described by
the type-family in the `HasPagination` type-class.

So, let's start with some imports and extensions to get this out of the way:

```hs
{-# LANGUAGE DataKinds             #-}
{-# LANGUAGE DeriveGeneric         #-}
{-# LANGUAGE FlexibleInstances     #-}
{-# LANGUAGE MultiParamTypeClasses #-}
{-# LANGUAGE TypeApplications      #-}
{-# LANGUAGE TypeFamilies          #-}
{-# LANGUAGE TypeOperators         #-}

import           Data.Aeson         (ToJSON, genericToJSON)
import           Data.Maybe         (fromMaybe)
import           Data.Proxy         (Proxy (..))
import           GHC.Generics       (Generic)
import           Servant            ((:>), GetPartialContent, Handler, Header, Headers, JSON, Server)
import           Servant.Pagination (HasPagination (..), PageHeaders, Range (..), Ranges, applyRange)

import qualified Data.Aeson               as Aeson
import qualified Network.Wai.Handler.Warp as Warp
import qualified Servant.Pagination       as Pagination
import qualified Servant
```


#### Declaring the Resource

Servant APIs are rather resource-oriented, and so is `servant-pagination`. This
guide shows a basic example working with `JSON` (as you could tell from the
import list already). To make the world a <span style='text-decoration:
line-through'>better</span> colored place, let's create an API to retrieve
colors -- with pagination.

```hs
data Color = Color
  { name :: String
  , rgb  :: [Int]
  , hex  :: String
  } deriving (Eq, Show, Generic)

instance ToJSON Color where
  toJSON =
    genericToJSON Aeson.defaultOptions

colors :: [Color]
colors =
  [ Color "Black" [0, 0, 0] "#000000"
  , Color "Blue" [0, 0, 255] "#0000ff"
  , Color "Green" [0, 128, 0] "#008000"
  , Color "Grey" [128, 128, 128] "#808080"
  , Color "Purple" [128, 0, 128] "#800080"
  , Color "Red" [255, 0, 0] "#ff0000"
  , Color "Yellow" [255, 255, 0] "#ffff00"
  ]
```

#### Declaring the Ranges

Now that we have defined our _resource_ (a.k.a `Color`), we are ready to declare a new `Range`
that will operate on a "name" field (genuinely named after the `name` fields from the `Color`
record).
For that, we need to tell `servant-pagination` two things:

- What is the type of the corresponding `Range` values
- How do we get one of these values from our resource

This is done via defining an instance of `HasPagination` as follows:

```hs
instance HasPagination Color "name" where
  type RangeType Color "name" = String
  getFieldValue _ = name
  -- getRangeOptions :: Proxy "name" -> Proxy Color -> RangeOptions
  -- getDefaultRange :: Proxy Color -> Range "name" String

defaultRange :: Range "name" String
defaultRange =
  getDefaultRange (Proxy @Color)
```

Note that `getFieldValue :: Proxy "name" -> Color -> String` is the minimal complete definintion
of the class. Yet, you can define `getRangeOptions` to provide different parsing options (see
the last section of this guide). In the meantime, we've also defined a `defaultRange` as it will
come in handy when defining our handler.

#### API

Good, we have a resource, we have a `Range` working on that resource, we can now declare our
API using other Servant combinators we already know:

```hs
type API =
  "colors"
    :> Header "Range" (Ranges '["name"] Color)
    :> GetPartialContent '[JSON] (Headers MyHeaders [Color])

type MyHeaders =
  Header "Total-Count" Int ': PageHeaders '["name"] Color
```

`PageHeaders` is a type alias provided by the library to declare the necessary response headers
we mentionned in introduction. Expanding the alias boils down to the following:

```hs
-- type MyHeaders
--   =  Header "Total-Count"   Int
--   :> Header "Accept-Ranges" (AcceptRanges '["name"])
--   :> Header "Content-Range" (ContentRange '["name"] Color)
--   :> Header "Next-Range"    (Ranges '["name"] Color)
```

As a result, we will need to provide all those headers with the response in our handler. Worry
not, _servant-pagination_ provides an easy way to lift a collection of resources into such handler.

#### Server

Time to connect the last bits by defining the server implementation of our colorful API. The `Ranges`
type we've defined above (tight to the `Range` HTTP header) indicates the server to parse any `Range`
header, looking for the format defined in introduction with fields and target types we have just declared.
If no such header is provided, we will end up receiving `Nothing`. Otherwise, it will be possible
to _extract_ a `Range` from our `Ranges`.

```hs
server :: Server API
server = handler
  where
    handler :: Maybe (Ranges '["name"] Color) -> Handler (Headers MyHeaders [Color])
    handler mrange = do
      let range =
            fromMaybe defaultRange (mrange >>= extractRange)

      addHeader (length colors) <$> returnRange range (applyRange range colors)

main :: IO ()
main =
  Warp.run 1442 $ Servant.serve (Proxy @API) server
```

Let's try it out using different ranges to observe the server's behavior. As a reminder, here's
the format we defined, where `<field>` here can only be `name` and `<value>` must parse to a `String`:

- `Range: <field> [<value>][; offset <o>][; limit <l>][; order <asc|desc>]`

Beside the target field, everything is pretty much optional in the `Range` HTTP header. Missing parts
are deducted from the `RangeOptions` that are part of the `HasPagination` instance. Therefore, all
following examples are valid requests to send to our server:

- 1 - `curl http://localhost:1442/colors -vH 'Range: name'`
- 2 - `curl http://localhost:1442/colors -vH 'Range: name; limit 2'`
- 3 - `curl http://localhost:1442/colors -vH 'Range: name Green; order asc; offset 1'`

Considering the following default options:

- `defaultRangeLimit: 100`
- `defaultRangeOffset: 0`
- `defaultRangeOrder: RangeDesc`

The previous ranges reads as follows:

- 1 - The first 100 colors, ordered by descending names
- 2 - The first 2 colors, ordered by descending names
- 3 - The 100 colors after `Green` (not included), ordered by ascending names.

> See `examples/Simple.hs` for a running version of this guide.

## Going Forward

####  Multiple Ranges

Note that in the simple above scenario, there's no ambiguity with `extractRange` and `returnRange`
because there's only one possible `Range` defined on our resource. Yet, as you've most probably
noticed, the `Ranges` combinator accepts a list of fields, each of which must declare a `HasPagination`
instance. Doing so will make the other helper functions more ambiguous and type annotation are
highly likely to be needed.


```hs
instance HasPagination Color "hex" where
  type RangeType Color "hex" = String
  getFieldValue _ = hex

-- to then define: Ranges '["name", "hex"] Color
```

> See `examples/Complex.hs` for more complex examples.


#### Parsing Options

By default, `servant-pagination` provides an implementation of `getRangeOptions` for each
`HasPagination` type-class. However, this can be overwritten when defining a instance of that
class to provide your own options. This options come into play when a `Range` header is
received and isn't fully specified (`limit`, `offset`, `order` are all optional) to provide
default fallback values for those.

For instance, let's say we wanted to change the default limit to `5` in a new range on
`"rgb"`, we could tweak the corresponding `HasPagination` instance as follows:

```hs
instance HasPagination Color "rgb" where
  type RangeType Color "rgb" = String
  getFieldValue _ = sum . rgb
  getRangeOptions _ _ = defaultOptions { defaultRangeLimit = 5 }
```

## Changelog

[CHANGELOG.md](CHANGELOG.md)


## License

[LGPL-3 © 2018 Chordify](LICENSE)


CHANGELOG.md
# Changelog

## Unreleased

- Move header logic to a pure addPageHeaders function #21

## v2.5.1 (2023-12-13)

- Bump `servant`, `servant-server`, `text` versions (see [#24](https://github.com/chordify/haskell-servant-pagination/pull/24)).

## v2.5.0 (2022-03-09)
- Allow text-2.0 and servant-0.19 #20
- Expose PutRange so we can use this in integrations #18

## v2.4.2 (2021-12-10)
- Don't return Next-Range upon partial response

## v2.4.1 (2021-10-11)
- Fix Travis CI

## v2.4.0 (2021-10-11)
- Update servant package bounds
- Forbid negative values for limit

## v2.3.0 (2020-03-05)

- Allow server-0.17 in executables
- Bump stack LTS to 14.25

## v2.2.2 (2019-02-28)

- (oversight) Allow server-0.16 in executables


## v2.2.1 (2019-02-28)

- Allow server-0.16


## v2.2.0 (2019-01-28)

- Use URL encoding for range values to properly support strings.


## v2.1.3 (2018-11-14)

- Fix Haddock generation


## v2.1.2 (2018-11-13)

- Allow servant-0.15


## v2.1.1 (2018-07-10)

- Allow servant-0.14


## v2.1.0 (2018-04-16)

- Add some tests (QuickCheck round-up & control some Ranges parsing)
- Add `Show` and `Eq` instances for Ranges
- Expose `putRange` function
- Review `getDefaultRange` signature (remove Maybe argument)


## v2.0.0 (2018-04-06)

- Review internal implementation and public API (ditch Range combinator to favor type-level
  list and more discrete footprint).

- Remove 'Total-Count' header, can still be added on top of the range headers but isn't a Range
  header so to speak.

- Extend haddock documentation to be more user-friendly


## v1.0.0 (2018-02-06)

- Initial release


examples/Simple.hs
{-# OPTIONS_GHC -fno-warn-orphans #-}
{-# LANGUAGE TypeApplications #-}
{-# LANGUAGE TypeFamilies     #-}

module Main where

import           Data.Maybe               (fromMaybe)
import           Data.Proxy               (Proxy (..))
import           Servant
import           Servant.Pagination

import qualified Network.Wai.Handler.Warp as Warp

import           Color


--  Ranges definitions

instance HasPagination Color "name" where
  type RangeType Color "name" = String
  getFieldValue _ = name

-- API

type API =
  "colors"
    :> Header "Range" (Ranges '["name"] Color)
    :> GetPartialContent '[JSON] (Headers (PageHeaders '["name"] Color) [Color])


-- Application

defaultRange :: Range "name" String
defaultRange =
  getDefaultRange (Proxy @Color)

server :: Server API
server mrange = do
  let range =
        fromMaybe defaultRange (mrange >>= extractRange)

  returnRange range (applyRange range colors)

main :: IO ()
main =
  Warp.run 1337 (serve (Proxy :: Proxy API) server)


-- Examples

-- $ curl -v http://localhost:1337/colors
--
-- > GET /colors HTTP/1.1
-- > Host: localhost:1337
-- > User-Agent: curl/7.47.0
-- > Accept: */*
-- >
-- < HTTP/1.1 206 Partial Content
-- < Transfer-Encoding: chunked
-- < Date: Tue, 30 Jan 2018 12:45:17 GMT
-- < Server: Warp/3.2.13
-- < Content-Type: application/json;charset=utf-8
-- < Accept-Ranges: name
-- < Content-Range: name Yellow..Aqua
-- < Next-Range: name Aqua;limit 100;offset 1;order desc


-- $ curl -v http://localhost:1337/colors --header 'Range: name; offset 59'
--
-- > GET /colors HTTP/1.1
-- > Range: name; offset 59
-- >
-- < HTTP/1.1 206 Partial Content
-- < Content-Type: application/json;charset=utf-8
-- < Accept-Ranges: name


src/Servant/Pagination.hs
{-# LANGUAGE RecordWildCards  #-}
{-# LANGUAGE TypeApplications #-}

-- | Opinionated Pagination Helpers for Servant APIs
--
--
-- Client can provide a `Range` header with their request with the following format
--
-- > Range: <field> [<value>][; offset <o>][; limit <l>][; order <asc|desc>]
--
-- Available ranges are declared using type-level list of accepted fields, bound to a given
-- resource and type using the 'HasPagination' type-class. The library provides unobtrusive
-- types and abstract away all the plumbing to hook that on an existing Servant API.
--
-- The 'IsRangeType' constraints summarize all constraints that must apply to a possible field
-- and heavily rely on the 'Web.Internal.FromHttpApiData' and 'Web.Internal.ToHttpApiData'.
--
-- > $ curl -v http://localhost:1337/colors -H 'Range: name; limit 10'
-- >
-- > > GET /colors HTTP/1.1
-- > > Host: localhost:1337
-- > > User-Agent: curl/7.47.0
-- > > Accept: */*
-- > >
-- > < HTTP/1.1 206 Partial Content
-- > < Transfer-Encoding: chunked
-- > < Date: Tue, 30 Jan 2018 12:45:17 GMT
-- > < Server: Warp/3.2.13
-- > < Content-Type: application/json;charset=utf-8
-- > < Accept-Ranges: name
-- > < Content-Range: name Yellow..Purple
-- > < Next-Range: name Purple;limit 10;offset 1;order desc
--
-- The 'Range' header is totally optional, but when provided, it indicates to the server what
-- parts of the collection is requested. As a reponse and in addition to the data, the server may
-- provide 3 headers to the client:
--
-- - @Accept-Ranges@: A comma-separated list of field upon which a range can be defined
-- - @Content-Range@: Actual range corresponding to the content being returned
-- - @Next-Range@: Indicate what should be the next `Range` header in order to retrieve the next range
--
-- This allows the client to work in a very _dumb_ mode where it simply consumes data from the server
-- using the value of the 'Next-Range' header to fetch each new batch of data. The 'Accept-Ranges'
-- comes in handy to self-document the API telling the client about the available filtering and sorting options
-- of a resource.
--
-- Here's a minimal example used to obtained the previous behavior; Most of the magic happens in the
-- 'returnRange' function which lift a collection of resources into a Servant handler, computing the
-- corresponding ranges from the range used to retrieve the resources.
--
-- @
-- -- Resource Type
--
-- data Color = Color
--   { name :: 'String'
--   , rgb  :: ['Int']
--   , hex  :: 'String'
--   } deriving ('Eq', 'Show', 'GHC.Generics.Generic')
--
-- colors :: [Color]
-- colors = [ {- ... -} ]
--
-- -- Ranges definitions
--
-- instance 'HasPagination' Color "name" where
--   type 'RangeType' Color "name" = 'String'
--   'getFieldValue' _ = name
--
--
-- -- API
--
-- type API =
--   "colors"
--     :> 'Servant.Header' \"Range\" ('Ranges' '["name"] Color)
--     :> 'Servant.GetPartialContent' '['Servant.JSON'] ('Servant.Headers' ('PageHeaders' '["name"] Color) [Color])
--
--
-- -- Application
--
-- defaultRange :: 'Range' "name" 'String'
-- defaultRange =
--   'getDefaultRange' ('Data.Proxy.Proxy' \@Color)
--
-- server :: 'Servant.Server.Server' API
-- server mrange = do
--   let range =
--         'Data.Maybe.fromMaybe' defaultRange (mrange >>= 'extractRange')
--
--   'returnRange' range ('applyRange' range colors)
--
-- main :: 'IO' ()
-- main =
--   'Network.Wai.Handler.Warp.run' 1337 ('Servant.Server.serve' ('Data.Proxy.Proxy' \@API) server)
-- @
module Servant.Pagination
  (
  -- * Types
  Ranges
  , Range(..)
  , RangeOrder(..)
  , AcceptRanges (..)
  , ContentRange (..)
  , PageHeaders
  , IsRangeType
  , PutRange
  , ExtractRange

  -- * Declare Ranges
  , HasPagination(..)
  , RangeOptions(..)
  , defaultOptions

  -- * Use Ranges
  , extractRange
  , putRange
  , addPageHeaders
  , returnRange
  , applyRange
  ) where

import           Data.List          (find, intercalate)
import           Data.Maybe         (listToMaybe)
import           Data.OpenApi       (ToParamSchema (..))
import           Data.String        (fromString)
import           Data.Text          (Text)
import           GHC.Generics       (Generic)
import           GHC.TypeLits       (KnownSymbol, Symbol, symbolVal)
import           Network.URI.Encode (decodeText, encodeText)
import           Servant

import qualified Data.List          as List
import qualified Data.OpenApi       as O
import qualified Data.Text          as Text
import qualified Safe


--
-- TYPES
--

-- | Helper type to define 'Header' with 'Description'.
type HeaderWithDescription name a description =
  Header' '[Description description, Optional, Strict] name a

-- | Set of constraints that must apply to every type target of a 'Range'
type IsRangeType a =
  ( Show a
  , Ord a
  , Eq a
  , FromHttpApiData a
  , ToHttpApiData a
  )

-- | A type to specify accepted Ranges via the @Range@ HTTP Header. For example:
--
-- @
-- type API = "resources"
--   :> 'Servant.Header' \"Range\" ('Ranges' '["created_at"] Resource)
--   :> 'Servant.Get' '['Servant.JSON'] [Resource]
-- @
data Ranges :: [Symbol] -> * -> * where
  Lift   :: Ranges fields resource -> Ranges (y ': fields) resource
  Ranges
    :: HasPagination resource field
    => Range field (RangeType resource field)
    -> Ranges (field ': fields) resource

instance (Show (Ranges '[] res)) where
  showsPrec _ _ = flip mappend "Ranges"

instance (Show (Ranges fields res)) => Show (Ranges (field ': fields) res) where
  showsPrec prec (Lift r)   s = showsPrec prec r s
  showsPrec prec (Ranges r) s =
    let
      inner = "Ranges@" ++ showsPrec 11 r s
    in
      if prec > 10 then "(" ++ inner ++ ")" else inner


-- | An actual 'Range' instance obtained from parsing / to generate a @Range@ HTTP Header.
data Range (field :: Symbol) (a :: *) =
  (KnownSymbol field, IsRangeType a) => Range
  { rangeValue  :: Maybe a     -- ^ The value of that field, beginning of the range
  , rangeLimit  :: Int         -- ^ Maximum number of resources to return
  , rangeOffset :: Int         -- ^ Offset, number of resources to skip after the starting value
  , rangeOrder  :: RangeOrder  -- ^ The order of sorting (ascending or descending)
  , rangeField  :: Proxy field -- ^ Actual field this Range actually refers to
  }

instance Eq (Range field a) where
  (Range val0 lim0 off0 ord0 _) == (Range val1 lim1 off1 ord1 _) =
       val0 == val1
    && lim0 == lim1
    && off0 == off1
    && ord0 == ord1

instance Show (Range field a) where
  showsPrec prec Range{..} =
    let
      inner = "Range {" ++ args ++ "}"
      args  = intercalate ", "
        [ "rangeValue = "  ++ show rangeValue
        , "rangeLimit = "  ++ show rangeLimit
        , "rangeOffset = " ++ show rangeOffset
        , "rangeOrder = "  ++ show rangeOrder
        , "rangeField = "  ++ "\"" ++ symbolVal rangeField ++ "\""
        ]
    in
      flip mappend $ if prec > 10 then
        "(" ++ inner ++ ")"
      else
        inner


-- | Extract a 'Range' from a 'Ranges'
class ExtractRange (fields :: [Symbol]) (field :: Symbol) where
  -- | Extract a 'Range' from a 'Ranges'. Works like a safe 'read', trying to coerce a 'Range' instance to
  -- an expected type. Type annotation are most likely necessary to remove ambiguity. Note that a 'Range'
  -- can only be extracted to a type bound by the allowed 'fields' on a given 'resource'.
  --
  -- @
  -- extractDateRange :: 'Ranges' '["created_at", "name"] Resource -> 'Range' "created_at" 'Data.Time.Clock.UTCTime'
  -- extractDateRange =
  --   'extractRange'
  -- @
  extractRange
    :: HasPagination resource field
    => Ranges fields resource                         -- ^ A list of accepted Ranges for the API
    -> Maybe (Range field (RangeType resource field)) -- ^ A Range instance of the expected type, if it matches

instance ExtractRange (field ': fields) field where
  extractRange (Ranges r) = Just r
  extractRange (Lift _)   = Nothing
  {-# INLINE extractRange #-}

instance {-# OVERLAPPABLE #-} ExtractRange fields field => ExtractRange (y ': fields) field where
  extractRange (Ranges _) = Nothing
  extractRange (Lift r)   = extractRange r
  {-# INLINE extractRange #-}


-- | Put a 'Range' in a 'Ranges'
class PutRange (fields :: [Symbol]) (field :: Symbol) where
  putRange
    :: HasPagination resource field
    => Range field (RangeType resource field)
    -> Ranges fields resource

instance PutRange (field ': fields) field where
  putRange = Ranges
  {-# INLINE putRange #-}

instance {-# OVERLAPPABLE #-} (PutRange fields field) => PutRange (y ': fields) field where
  putRange = Lift . putRange
  {-# INLINE putRange #-}

instance ToHttpApiData (Ranges fields resource) where
  toUrlPiece (Lift range) =
    toUrlPiece range

  toUrlPiece (Ranges Range{..}) =
    Text.pack (symbolVal rangeField)
    <> maybe "" (\v -> " " <> (encodeText . toUrlPiece) v) rangeValue
    <> ";limit "  <> toUrlPiece rangeLimit
    <> ";offset " <> toUrlPiece rangeOffset
    <> ";order "  <> toUrlPiece rangeOrder


instance FromHttpApiData (Ranges '[] resource) where
  parseUrlPiece _ =
    Left "Invalid Range"

instance
  ( FromHttpApiData (Ranges fields resource)
  , HasPagination resource field
  , KnownSymbol field
  , IsRangeType (RangeType resource field)
  ) => FromHttpApiData (Ranges (field ': fields) resource) where
  parseUrlPiece txt =
    let
      RangeOptions{..} = getRangeOptions (Proxy @field) (Proxy @resource)

      toTuples =
        filter (/= "") . Text.splitOn (Text.singleton ' ')

      args =
        map toTuples $ Text.splitOn (Text.singleton ';') txt

      field =
        Text.pack $ symbolVal (Proxy @field)
    in
      case args of
        (field' : value) : rest | field == field' -> do
          opts <-
            traverse parseOpt rest

          range <- Range
            <$> mapM (parseUrlPiece . decodeText) (listToMaybe value)
            <*> (ifOpt "limit"  defaultRangeLimit opts >>= checkLimit)
            <*> ifOpt "offset" defaultRangeOffset opts
            <*> ifOpt "order"  defaultRangeOrder opts
            <*> pure (Proxy @field)

          pure $ Ranges range

        _ ->
          Lift <$> (parseUrlPiece txt :: Either Text (Ranges fields resource))
    where
      parseOpt :: [Text] -> Either Text (Text, Text)
      parseOpt piece =
        case piece of
          [opt, arg] ->
            pure (opt, arg)

          _ ->
            Left "Invalid Range Options"

      ifOpt :: FromHttpApiData o => Text -> o -> [(Text, Text)] -> Either Text o
      ifOpt opt def =
        maybe (pure def) (parseQueryParam . snd) . find ((== opt) . fst)

      checkLimit :: Int -> Either Text Int
      checkLimit n
        | n < 0     = Left "Limit must be non-negative"
        | otherwise = return n

instance ToParamSchema (Ranges fields resource) where
  toParamSchema _ =
    mempty
      { O._schemaType = Just O.OpenApiString
      , O._schemaFormat =
          Just "<field> [<value>][; offset <o>][; limit <l>][; order <asc|desc>]"
      , O._schemaExample =
          Just $ fromString "createdAt 2017-02-19T12%3A56%3A28.000Z; offset 0; limit 100; order desc"
      }

-- | Define the sorting order of the paginated resources (ascending or descending)
data RangeOrder
  = RangeAsc
  | RangeDesc
  deriving (Eq, Show, Ord, Generic)

instance ToHttpApiData RangeOrder where
  toUrlPiece order =
    case order of
      RangeAsc  -> "asc"
      RangeDesc -> "desc"

instance FromHttpApiData RangeOrder where
  parseUrlPiece txt =
    case txt of
      "asc"  -> pure RangeAsc
      "desc" -> pure RangeDesc
      _      -> Left "Invalid Range Order"


-- | Type alias to declare response headers related to pagination
--
-- @
-- type MyHeaders =
--   'PageHeaders' '["created_at"] Resource
--
-- type API = "resources"
--   :> 'Servant.Header' \"Range\" ('Ranges' '["created_at"] Resource)
--   :> 'Servant.Get' '['Servant.JSON'] ('Servant.Headers' MyHeaders [Resource])
-- @
type PageHeaders (fields :: [Symbol]) (resource :: *) =
  '[ HeaderWithDescription "Accept-Ranges" (AcceptRanges fields)
      "A comma-separated list of fields upon which a pagination range can be defined"
   , HeaderWithDescription "Content-Range" (ContentRange fields resource)
      "Actual pagination range corresponding to the content being returned."
   , HeaderWithDescription "Next-Range"    (Ranges fields resource)
      "Indicate what should be the next Range header in order to retrieve the next range"
   ]

-- | Accepted Ranges in the `Accept-Ranges` response's header
data AcceptRanges (fields :: [Symbol]) = AcceptRanges
  deriving (Show, Eq)

instance ToHttpApiData (AcceptRanges '[]) where
  toUrlPiece AcceptRanges = mempty

instance (KnownSymbol field) => ToHttpApiData (AcceptRanges '[field]) where
  toUrlPiece AcceptRanges =
    Text.pack (symbolVal (Proxy @field))

instance (ToHttpApiData (AcceptRanges (f ': fs)), KnownSymbol field) => ToHttpApiData (AcceptRanges (field ': f ': fs)) where
  toUrlPiece AcceptRanges =
    Text.pack (symbolVal (Proxy @field)) <> "," <> toUrlPiece (AcceptRanges @(f ': fs))

instance FromHttpApiData (AcceptRanges '[]) where
  parseUrlPiece _ = Left "Invalid Accept Ranges"

instance KnownSymbol field => FromHttpApiData (AcceptRanges (field ': fields)) where
  parseUrlPiece txt =
    let field = Text.pack $ symbolVal (Proxy @field)
    in
      case Text.splitOn "," txt of
        field' : _ | field == field' -> pure $ AcceptRanges @(field ': fields)
        _ -> Left "Invalid Accept Ranges"

instance ToParamSchema (AcceptRanges fields) where
  toParamSchema _ =
    mempty
      { O._schemaType = Just O.OpenApiString
      , O._schemaExample = Just $ fromString "createdAt, modifiedAt"
      }

-- | Actual range returned, in the `Content-Range` response's header
data ContentRange (fields :: [Symbol]) resource =
  forall field. (KnownSymbol field, ToHttpApiData (RangeType resource field)) => ContentRange
  { contentRangeStart :: RangeType resource field
  , contentRangeEnd   :: RangeType resource field
  , contentRangeField :: Proxy field
  }

instance Eq (ContentRange (field ': fields) a) where
  (ContentRange start0 end0 _) == (ContentRange start1 end1 _) =
       toUrlPiece start0 == toUrlPiece start1
    && toUrlPiece end0 == toUrlPiece end1

instance Show (ContentRange (field ': fields) a) where
  showsPrec prec ContentRange{..} =
    let
      inner = "ContentRange {" ++ args ++ "}"
      args  = intercalate ", "
        [ "contentRangeStart = "  ++ Text.unpack (encodeText $ toUrlPiece contentRangeStart)
        , "contentRangeEnd   = "  ++ Text.unpack (encodeText $ toUrlPiece contentRangeEnd)
        , "contentRangeField = "  ++ "\"" ++ symbolVal contentRangeField ++ "\""
        ]
    in
      flip mappend $ if prec > 10 then
        "(" ++ inner ++ ")"
      else
        inner

instance ToHttpApiData (ContentRange fields res) where
  toUrlPiece (ContentRange start end field) =
    Text.pack (symbolVal field) <> " " <> (encodeText . toUrlPiece) start <> ".." <> (encodeText . toUrlPiece) end

instance FromHttpApiData (ContentRange '[] resource) where
  parseUrlPiece _ = Left "Invalid Content Range"

instance
  ( KnownSymbol field
  , ToHttpApiData (RangeType resource field)
  , FromHttpApiData (RangeType resource field)
  ) => FromHttpApiData (ContentRange (field ': fields) resource) where
  parseUrlPiece txt =
    case Text.splitOn ".." . snd $ Text.breakOnEnd " " txt of
      [start, end] ->
        ContentRange
          <$> parseUrlPiece start
          <*> parseUrlPiece end
          <*> pure (Proxy @field)
      _otherwise -> Left "Invalid Content Range"

instance ToParamSchema (ContentRange fields resource) where
  toParamSchema _ =
    mempty
      { O._schemaType = Just O.OpenApiString
      , O._schemaExample =
          Just $ fromString "createdAt 2017-01-15T23%3A14%3A51.000Z..2017-02-18T06%3A10%3A23.000Z"
      }

--
-- USE RANGES
--

-- | Available 'Range' on a given @resource@ must implements the 'HasPagination' type-class.
-- This class defines how the library can interact with a given @resource@ to access the value
-- to which a @field@ refers.
class KnownSymbol field => HasPagination resource field where
  type RangeType resource field :: *

  -- | Get the corressponding value of a Resource
  getFieldValue :: Proxy field -> resource -> RangeType resource field

  -- | Get parsing options for the 'Range' defined on this 'field'
  getRangeOptions :: Proxy field -> Proxy resource -> RangeOptions
  getRangeOptions _ _ = defaultOptions

  -- | Create a default 'Range' from a value and default 'RangeOptions'. Typical use-case
  -- is for when no or an invalid 'Range' header was provided.
  getDefaultRange
    :: IsRangeType (RangeType resource field)
    => Proxy resource
    -> Range field (RangeType resource field)
  getDefaultRange _ =
    let
      RangeOptions{..} = getRangeOptions (Proxy @field) (Proxy @resource)
    in Range
      { rangeValue  = Nothing @(RangeType resource field)
      , rangeLimit  = defaultRangeLimit
      , rangeOffset = defaultRangeOffset
      , rangeOrder  = defaultRangeOrder
      , rangeField  = Proxy @field
      }

-- | Add headers representing a 'Range' to a list of resources.
--
-- 'Ranges' headers can be quite cumbersome to declare and can be deduced from a
-- collection of resources together with the 'Range' used to retrieve it, so this function
-- is a shortcut for that.
--
-- @
-- myHandler
--  :: 'Maybe' ('Ranges' '["created_at"] Resource)
--  -> 'Servant.Server.Handler' ('Servant.Headers' ('PageHeaders' '["created_at"] Resource) [Resource])
-- myHandler mrange =
--  let range =
--        'Data.Maybe.fromMaybe' ('getDefaultRange' ('Data.Proxy.Proxy' \@Resource)) (mrange >>= 'extractRange')
--
--  'return' ('addPageHeaders' range ('applyRange' range resources))
-- @
addPageHeaders
  :: ( ToHttpApiData (AcceptRanges fields)
     , KnownSymbol field
     , HasPagination resource field
     , IsRangeType (RangeType resource field)
     , PutRange fields field
     )
  => Range field (RangeType resource field)               -- ^ Actual 'Range' used to retrieve the results
  -> [resource]                                           -- ^ Resources to return, fetched from a db or a local store
  -> Headers (PageHeaders fields resource) [resource]     -- ^ The same resources, but with pagination headers
addPageHeaders Range{..} rs =
  let boundaries = (,)
        <$> fmap (getFieldValue rangeField) (Safe.headMay rs)
        <*> fmap (getFieldValue rangeField) (Safe.lastMay rs)
  in case boundaries of
    Nothing ->
      addHeader AcceptRanges $ noHeader $ noHeader rs

    Just (start, end) -> do
      let nextOffset | rangeValue == Just end = rangeOffset + length rs
                     | otherwise              = length $ takeWhile ((==) end . getFieldValue rangeField) (reverse rs)

      let nextRange = putRange Range
            { rangeValue  = Just end
            , rangeLimit  = rangeLimit
            , rangeOffset = nextOffset
            , rangeOrder  = rangeOrder
            , rangeField  = rangeField
            }

      let contentRange = ContentRange
            { contentRangeStart = start
            , contentRangeEnd   = end
            , contentRangeField = rangeField
            }

      let addNextRange | length rs < rangeLimit = noHeader
                       | otherwise              = addHeader nextRange

      addHeader AcceptRanges $ addHeader contentRange $ addNextRange rs

-- | @'returnRange' range rs = 'return' ('addPageHeaders' range rs)@
returnRange
  :: ( Monad m
     , ToHttpApiData (AcceptRanges fields)
     , KnownSymbol field
     , HasPagination resource field
     , IsRangeType (RangeType resource field)
     , PutRange fields field
     )
  => Range field (RangeType resource field)               -- ^ Actual 'Range' used to retrieve the results
  -> [resource]                                           -- ^ Resources to return, fetched from a db or a local store
  -> m (Headers (PageHeaders fields resource) [resource]) -- ^ Resources embedded in a given 'Monad' (typically a 'Servant.Server.Handler', with pagination headers)
returnRange range rs = return (addPageHeaders range rs)

-- | Default values to apply when parsing a 'Range'
data RangeOptions  = RangeOptions
  { defaultRangeLimit  :: Int         -- ^ Default limit if not provided, default to @100@
  , defaultRangeOffset :: Int         -- ^ Default offset if not provided, default to @0@
  , defaultRangeOrder  :: RangeOrder  -- ^ Default order if not provided, default to 'RangeDesc'
  } deriving (Eq, Show)


-- | Some default options of default values for a Range (limit 100; offset 0; order desc)
defaultOptions :: RangeOptions
defaultOptions =
  RangeOptions 100 0 RangeDesc


-- | Helper to apply a 'Range' to a list of values. Most likely useless in practice
-- as results may come more realistically from a database, but useful for debugging or
-- testing.
applyRange
  :: HasPagination resource field
  => Range field (RangeType resource field) -- ^ A 'Range' instance on a given @resource@
  -> [resource]                             -- ^ A full-list of @resource@
  -> [resource]                             -- ^ The sublist obtained by applying the 'Range'
applyRange Range{..} =
  let
    sortRel =
      case rangeOrder of
        RangeDesc ->
          \a b -> compare (getFieldValue rangeField b) (getFieldValue rangeField a)

        RangeAsc ->
          \a b -> compare (getFieldValue rangeField a) (getFieldValue rangeField b)

    dropRel =
      case (rangeValue, rangeOrder) of
        (Nothing, _) ->
          const False

        (Just a, RangeDesc) ->
          (> a) . getFieldValue rangeField

        (Just a, RangeAsc) ->
          (< a) . getFieldValue rangeField
  in
      List.take rangeLimit
    . List.drop rangeOffset
    . List.dropWhile dropRel
    . List.sortBy sortRel
